### PR TITLE
[SPARK-41132][SQL] Convert LikeAny and NotLikeAny to InSet if no pattern contains wildcards

### PR DESCRIPTION
**Pull request by wankunde**
### What changes were proposed in this pull request?

Improve likeAny and notLikeAny performance.

We can optimize query `SELECT * FROM tab WHERE trim(addr) LIKE ANY ('5001', '5002%', '%5003', '5004', '5001')` to `SELECT * FROM tab WHERE trim(addr) IN ('5001', '5004') OR trim(addr) like '5002%' OR trim(addr) like '%5003'`

Before this PR
```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_281-b09 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Multi like query:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Query with LikeAny simplification                  2335           2456         184          0.0      233496.6       1.0X
```
After this PR
```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_281-b09 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Multi like query:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Query with LikeAny simplification                  1912           1966          50          0.0      191230.9       1.0X
```

### Why are the changes needed?

Match with set values will be faster than regex expressions.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Add UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 1]
```diff
@@ -780,6 +780,13 @@ object LikeSimplification extends Rule[LogicalPlan] {
       } else {
         simplifyLike(input, pattern.toString, escapeChar).getOrElse(l)
       }
+    case LikeAny(child, patterns)
+      if patterns.map(_.toString).forall { case equalTo(_) => true case _ => false } =>
+      InSet(child, patterns.toSet)
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:785]: Could we also support this case?
```sql
SELECT * FROM tab WHERE trim(addr) LIKE ANY ('5000', '5001', '5002%', '5003%', '5004')
```

Comment (wankunde) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Now support this pattern, and add UT

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 2]
```diff
@@ -762,10 +762,40 @@ object LikeSimplification extends Rule[LogicalPlan] with PredicateHelper {
           val and = buildBalancedPredicate(replacements.map(Not(_)), And)
           if (remainPatterns.nonEmpty) And(and, l.copy(patterns = remainPatterns)) else and
         case l: LikeAny =>
-          val or = buildBalancedPredicate(replacements, Or)
+          val equalPatterns = MutableHashset[Any]()
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:765]: Do you have some benchmark number?

Comment (wankunde) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:765]: Added benchmark result.

benchmark code:
```
object LikeAnyBenchmark extends SqlBasedBenchmark {
  import spark.implicits._

  private val numRows = 10000
  private val width = 5

  def withTempTable(tableNames: String*)(f: => Unit): Unit = {
    try f finally tableNames.foreach(spark.catalog.dropTempView)
  }

  private def saveAsTable(df: DataFrame, dir: File): Unit = {
    val parquetPath = dir.getCanonicalPath + "/parquet"
    df.write.mode("overwrite").parquet(parquetPath)
    spark.read.parquet(parquetPath).createOrReplaceTempView("parquetTable")
  }

  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
    withTempPath { dir =>
      withTempTable("parquetTable") {
        val selectExpr = (1 to width).map(i => s"CAST(value + 1000000 AS STRING) c$i")
        val df = spark.range(0, numRows, 1, 100)
          .map(_ => Random.nextLong).selectExpr(selectExpr: _*)
        saveAsTable(df, dir)

        val benchmark =
          new Benchmark("Multi like query", numRows, minNumIters = 3, output = output)

        benchmark.addCase("Query with LikeAny simplification", numIters = 3) { _ =>
          val likeAnyExpr =
            Random.shuffle(Range(1000, 1300).map(i =>
              if (i < 1100) s"'$i%'" else if (i < 1200) s"'%$i'" else s"'$i'"
            )).mkString("c1 like any(", ", ", ")")
          spark.sql(s"SELECT * FROM parquetTable WHERE $likeAnyExpr").noop()
        }
        benchmark.run()
      }
    }
  }
}
```

**Pull request closed without merging**